File: bdd/java/src/test/java/org/apache/iggy/bdd/LeaderRedirectionSteps.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.iggy.bdd;

import io.cucumber.java.After;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import org.apache.iggy.client.ConnectionInfo;
import org.apache.iggy.client.blocking.tcp.IggyTcpClient;
import org.apache.iggy.cluster.ClusterNode;
import org.apache.iggy.cluster.ClusterNodeRole;
import org.apache.iggy.cluster.ClusterNodeStatus;
import org.apache.iggy.exception.IggyException;
import org.apache.iggy.stream.StreamDetails;

import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class LeaderRedirectionSteps {

    private static final String MAIN_CLIENT = "main";

    private final Map<String, IggyTcpClient> clients = new LinkedHashMap<>();
    private boolean redirectionOccurred;
    private Long lastStreamId;

    @After
    public void closeClients() {
        clients.values().forEach(IggyTcpClient::close);
        clients.clear();
    }

    // ---------- Background ----------

    @Given("I have cluster configuration enabled with {int} nodes")
    public void clusterConfigurationEnabled(int nodeCount) {
        assertTrue(nodeCount > 0, "Cluster must have at least one node");
    }

    @Given("node {int} is configured on port {int}")
    public void nodeConfiguredOnPort(int nodeId, int port) {
        assertFalse(addressForPort(port).isBlank(), "Node " + nodeId + " address should be configured");
    }

    // ---------- Server start steps (validate addresses are configured) ----------

    @Given("^I start server (\\d+) on port (\\d+) as (leader|follower)$")
    public void startServerAs(int nodeId, int port, String role) {
        assertAddressMatchesPort(addressForRole(role), port, role + " address");
    }

    @Given("I start a single server on port {int} without clustering enabled")
    public void startSingleServer(int port) {
        assertAddressMatchesPort(singleServerAddress(), port, "single-server address");
    }

    // ---------- Client creation ----------

    @When("^I create a client connecting to (follower|leader) on port (\\d+)$")
    public void createClientConnectingToRole(String role, int port) {
        String address = addressForRole(role);
        assertAddressMatchesPort(address, port, role + " address");
        createAndConnectClient(MAIN_CLIENT, address);
    }

    @When("I create a client connecting directly to leader on port {int}")
    public void createClientConnectingDirectlyToLeader(int port) {
        String address = addressForRole("leader");
        assertAddressMatchesPort(address, port, "leader address");
        createAndConnectClient(MAIN_CLIENT, address);
        redirectionOccurred = false;
    }

    @When("I create a client connecting to port {int}")
    public void createClientConnectingToPort(int port) {
        createAndConnectClient(MAIN_CLIENT, addressForPort(port));
    }

    @When("^I create client ([A-Z]) connecting to port (\\d+)$")
    public void createNamedClientConnectingToPort(String clientName, int port) {
        createAndConnectClient(clientName, addressForPort(port));
    }

    // ---------- Auth ----------

    @When("I authenticate as root user")
    public void authenticateAsRootUser() {
        authenticateAllClients();
    }

    @When("both clients authenticate as root user")
    public void bothClientsAuthenticateAsRootUser() {
        authenticateAllClients();
    }

    // ---------- Stream operations ----------

    @When("I create a stream named {string}")
    public void createStreamNamed(String streamName) {
        StreamDetails stream = client(MAIN_CLIENT).streams().createStream(streamName);
        assertNotNull(stream, "Stream should be created");
        lastStreamId = stream.id();
    }

    @Then("the stream should be created successfully on the leader")
    public void streamCreatedSuccessfullyOnLeader() {
        assertNotNull(lastStreamId, "Stream should have been created on the leader");
    }

    // ---------- Connection / redirection assertions ----------

    @Then("the client should automatically redirect to leader on port {int}")
    public void clientAutomaticallyRedirectsToLeader(int expectedPort) {
        verifyClientPort(MAIN_CLIENT, expectedPort);
    }

    @Then("^client ([A-Z]) should stay connected to port (\\d+)$")
    public void namedClientStaysConnectedToPort(String clientName, int expectedPort) {
        verifyClientPort(clientName, expectedPort);
    }

    @Then("^client ([A-Z]) should redirect to port (\\d+)$")
    public void namedClientRedirectsToPort(String clientName, int expectedPort) {
        verifyClientPort(clientName, expectedPort);
    }

    @Then("the client should not perform any redirection")
    public void clientDoesNotRedirect() {
        assertFalse(redirectionOccurred, "No redirection should occur when connecting directly to the leader");
    }

    @Then("the connection should remain on port {int}")
    public void connectionRemainsOnPort(int expectedPort) {
        verifyClientPort(MAIN_CLIENT, expectedPort);
        assertFalse(redirectionOccurred, "Connection should not have been redirected");
    }

    @Then("the client should connect successfully without redirection")
    public void clientConnectsWithoutRedirection() {
        client(MAIN_CLIENT).system().ping();
        assertFalse(redirectionOccurred, "No redirection should occur without clustering");
    }

    @Then("both clients should be using the same server")
    public void bothClientsUseTheSameServer() {
        IggyTcpClient clientA = client("A");
        IggyTcpClient clientB = client("B");
        ConnectionInfo connectionA = clientA.getConnectionInfo();
        ConnectionInfo connectionB = clientB.getConnectionInfo();

        assertTrue(
                isSameEndpoint(connectionA, connectionB),
                () -> "Both clients should be connected to the same server, got "
                        + connectionA.serverAddress()
                        + " and "
                        + connectionB.serverAddress());

        clientA.system().ping();
        clientB.system().ping();

        Optional<ClusterNode> leaderA = leaderFromMetadata(clientA);
        Optional<ClusterNode> leaderB = leaderFromMetadata(clientB);
        if (leaderA.isPresent() && leaderB.isPresent()) {
            assertEquals(
                    leaderA.get().ip() + ":" + leaderA.get().endpoints().tcp(),
                    leaderB.get().ip() + ":" + leaderB.get().endpoints().tcp(),
                    "Both clients should see the same leader");
        }
    }

    // ---------- Helpers ----------

    private void createAndConnectClient(String name, String address) {
        String[] hostPort = address.split(":", 2);
        IggyTcpClient client = IggyTcpClient.builder()
                .host(hostPort[0])
                .port(Integer.parseInt(hostPort[1]))
                .build();
        client.connect();
        clients.put(name, client);
    }

    private void authenticateAllClients() {
        String username = getenvOrDefault("IGGY_ROOT_USERNAME", "iggy");
        String password = getenvOrDefault("IGGY_ROOT_PASSWORD", "iggy");
        for (IggyTcpClient client : clients.values()) {
            String initialAddress = client.getConnectionInfo().serverAddress();
            client.users().login(username, password);
            if (!initialAddress.equals(client.getConnectionInfo().serverAddress())) {
                redirectionOccurred = true;
            }
        }
    }

    private void verifyClientPort(String clientName, int expectedPort) {
        IggyTcpClient client = client(clientName);
        assertEquals(
                expectedPort,
                client.getConnectionInfo().port(),
                "Client " + clientName + " should be connected to port " + expectedPort);
        client.system().ping();
    }

    private IggyTcpClient client(String name) {
        IggyTcpClient client = clients.get(name);
        if (client == null) {
            throw new IllegalStateException("Client " + name + " should exist");
        }
        return client;
    }

    /**
     * The leader per the roster, or empty when clustering is unavailable
     * (a server too old for the command, or metadata otherwise rejected).
     */
    private static Optional<ClusterNode> leaderFromMetadata(IggyTcpClient client) {
        try {
            return client.system().getClusterMetadata().nodes().stream()
                    .filter(node -> node.role() == ClusterNodeRole.Leader && node.status() == ClusterNodeStatus.Healthy)
                    .findFirst();
        } catch (IggyException error) {
            return Optional.empty();
        }
    }

    private static boolean isSameEndpoint(ConnectionInfo left, ConnectionInfo right) {
        if (left.port() != right.port()) {
            return false;
        }

        InetAddress[] leftAddresses = resolveHost(left.host());
        InetAddress[] rightAddresses = resolveHost(right.host());
        return Arrays.stream(leftAddresses)
                .anyMatch(leftAddress -> Arrays.stream(rightAddresses).anyMatch(leftAddress::equals));
    }

    private static InetAddress[] resolveHost(String host) {
        try {
            return InetAddress.getAllByName(host);
        } catch (UnknownHostException error) {
            throw new AssertionError("Failed to resolve server host " + host, error);
        }
    }

    private static void assertAddressMatchesPort(String address, int port, String description) {
        assertTrue(address.endsWith(":" + port), description + " " + address + " should use port " + port);
    }

    private static String addressForRole(String role) {
        return switch (role) {
            case "leader" -> getenvOrDefault("IGGY_TCP_ADDRESS_LEADER", "127.0.0.1:8091");
            case "follower" -> getenvOrDefault("IGGY_TCP_ADDRESS_FOLLOWER", "127.0.0.1:8092");
            default -> throw new IllegalArgumentException("Unknown role: " + role);
        };
    }

    private static String addressForPort(int port) {
        return switch (port) {
            case 8090 -> singleServerAddress();
            case 8091 -> addressForRole("leader");
            case 8092 -> addressForRole("follower");
            default -> throw new IllegalArgumentException("Unknown port: " + port);
        };
    }

    private static String singleServerAddress() {
        return getenvOrDefault("IGGY_TCP_ADDRESS", "127.0.0.1:8090");
    }

    private static String getenvOrDefault(String key, String defaultValue) {
        String value = System.getenv(key);
        return value == null || value.isBlank() ? defaultValue : value;
    }
}

File: foreign/csharp/Iggy_SDK.Tests.BDD/StepDefinitions/LeaderRedirectionSteps.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System.Net;
using Apache.Iggy.Configuration;
using Apache.Iggy.Contracts;
using Apache.Iggy.Enums;
using Apache.Iggy.Exceptions;
using Apache.Iggy.Factory;
using Apache.Iggy.IggyClient;
using Reqnroll;
using Shouldly;
using TestContext = Apache.Iggy.Tests.BDD.Context.TestContext;

namespace Apache.Iggy.Tests.BDD.StepDefinitions;

[Binding]
public class LeaderRedirectionSteps
{
    private const string RootUsername = "iggy";
    private const string RootPassword = "iggy";

    private readonly TestContext _context;

    public LeaderRedirectionSteps(TestContext context)
    {
        _context = context;
    }

    // ---------- Background ----------

    [Given(@"I have cluster configuration enabled with (\d+) nodes")]
    public void GivenIHaveClusterConfigurationEnabledWithNodes(int nodeCount)
    {
        nodeCount.ShouldBeGreaterThan(0);
    }

    [Given(@"node (\d+) is configured on port (\d+)")]
    public void GivenNodeIsConfiguredOnPort(int nodeId, int port)
    {
        ResolveAddressForPort(port).ShouldNotBeNullOrEmpty();
    }

    // ---------- Server start steps (just validate addresses are configured) ----------

    [Given(@"I start server (\d+) on port (\d+) as (leader|follower)")]
    public void GivenIStartServerOnPortAs(int nodeId, int port, string role)
    {
        var address = ResolveAddressForRole(role);
        if (!address.EndsWith($":{port}"))
        {
            throw new InvalidOperationException(
                $"{role} address {address} does not match expected port {port}");
        }
    }

    [Given(@"I start a single server on port (\d+) without clustering enabled")]
    public void GivenIStartASingleServerOnPortWithoutClusteringEnabled(int port)
    {
        if (!_context.TcpUrl.EndsWith($":{port}"))
        {
            throw new InvalidOperationException(
                $"Single-server address {_context.TcpUrl} does not match expected port {port}");
        }
    }

    // ---------- Client creation ----------

    [When(@"I create a client connecting to (follower|leader) on port (\d+)")]
    public async Task WhenICreateAClientConnectingToOnPort(string role, int port)
    {
        var address = ResolveAddressForRole(role);
        if (!address.EndsWith($":{port}"))
        {
            throw new InvalidOperationException(
                $"{role} address {address} does not match expected port {port}");
        }
        await CreateAndConnectClient("main", address);
    }

    [When(@"I create a client connecting directly to leader on port (\d+)")]
    public async Task WhenICreateAClientConnectingDirectlyToLeaderOnPort(int port)
    {
        var address = _context.LeaderTcpUrl;
        if (!address.EndsWith($":{port}"))
        {
            throw new InvalidOperationException(
                $"Leader address {address} does not match expected port {port}");
        }
        await CreateAndConnectClient("main", address);
        _context.RedirectionOccurred = false;
    }

    [When(@"I create a client connecting to port (\d+)")]
    public async Task WhenICreateAClientConnectingToPort(int port)
    {
        await CreateAndConnectClient("main", ResolveAddressForPort(port));
    }

    [When(@"I create client ([A-Z]) connecting to port (\d+)")]
    public async Task WhenICreateNamedClientConnectingToPort(string clientName, int port)
    {
        await CreateAndConnectClient(clientName, ResolveAddressForPort(port));
    }

    // ---------- Auth ----------

    [When(@"I authenticate as root user")]
    public async Task WhenIAuthenticateAsRootUser()
    {
        await AuthenticateAllClients();
    }

    [When(@"both clients authenticate as root user")]
    public async Task WhenBothClientsAuthenticateAsRootUser()
    {
        await AuthenticateAllClients();
    }

    private async Task AuthenticateAllClients()
    {
        var names = _context.Clients.Count > 1
            ? _context.Clients.Keys.ToList()
            : new List<string> { "main" };

        foreach (var name in names)
        {
            var client = GetClient(name);
            var initialAddress = client.GetCurrentAddress();
            var result = await client.LoginUserAsync(RootUsername, RootPassword);
            result.ShouldNotBeNull("Failed to login as root");

            // RedirectAsync runs inside LoginUserAsync; if address changed, redirection happened.
            if (client.GetCurrentAddress() != initialAddress)
            {
                _context.RedirectionOccurred = true;
            }

            if (_context.Clients.Count > 1)
            {
                await Task.Delay(100);
            }
        }
    }

    // ---------- Stream operations ----------

    [When(@"I create a stream named ""(.+)""")]
    public async Task WhenICreateAStreamNamed(string streamName)
    {
        var client = GetClient("main");
        var stream = await client.CreateStreamAsync(streamName);
        stream.ShouldNotBeNull("Should be able to create stream");
        _context.LastStreamId = stream.Id;
    }

    [Then(@"the stream should be created successfully on the leader")]
    public void ThenTheStreamShouldBeCreatedSuccessfullyOnTheLeader()
    {
        _context.LastStreamId.ShouldNotBeNull("Stream should have been created on leader");
    }

    // ---------- Connection / redirection assertions ----------

    [Then(@"the client should automatically redirect to leader on port (\d+)")]
    public async Task ThenTheClientShouldAutomaticallyRedirectToLeaderOnPort(int expectedPort)
    {
        await VerifyClientPort("main", expectedPort, markRedirection: true);
    }

    [Then(@"client ([A-Z]) should stay connected to port (\d+)")]
    public async Task ThenNamedClientShouldStayConnectedToPort(string clientName, int expectedPort)
    {
        await VerifyClientPort(clientName, expectedPort, markRedirection: false);
    }

    [Then(@"client ([A-Z]) should redirect to port (\d+)")]
    public async Task ThenNamedClientShouldRedirectToPort(string clientName, int expectedPort)
    {
        await VerifyClientPort(clientName, expectedPort, markRedirection: true);
    }

    [Then(@"the client should not perform any redirection")]
    public void ThenTheClientShouldNotPerformAnyRedirection()
    {
        _context.RedirectionOccurred.ShouldBeFalse(
            "No redirection should occur when connecting directly to leader");
    }

    [Then(@"the connection should remain on port (\d+)")]
    public async Task ThenTheConnectionShouldRemainOnPort(int port)
    {
        var client = GetClient("main");
        await VerifyClientConnection(client, port);
        _context.RedirectionOccurred.ShouldBeFalse("Connection should not have been redirected");
    }

    [Then(@"the client should connect successfully without redirection")]
    public async Task ThenTheClientShouldConnectSuccessfullyWithoutRedirection()
    {
        var client = GetClient("main");
        await client.PingAsync();
        _context.RedirectionOccurred.ShouldBeFalse("No redirection should occur without clustering");
    }

    [Then(@"both clients should be using the same server")]
    public async Task ThenBothClientsShouldBeUsingTheSameServer()
    {
        var clientA = GetClient("A");
        var clientB = GetClient("B");

        // A client that never redirected still holds the address it was given (possibly a
        // hostname), while a redirected one holds the roster address the cluster published.
        // Both may name the same node, so they are compared once resolved.
        await AssertSameEndpointAsync(clientA.GetCurrentAddress(), clientB.GetCurrentAddress());

        await clientA.PingAsync();
        await clientB.PingAsync();

        var leaderA = await GetLeaderFromMetadata(clientA);
        var leaderB = await GetLeaderFromMetadata(clientB);
        if (leaderA is not null && leaderB is not null)
        {
            $"{leaderA.Ip}:{leaderA.Endpoints.Tcp}".ShouldBe(
                $"{leaderB.Ip}:{leaderB.Endpoints.Tcp}",
                "Both clients should see the same leader");
        }
    }

    // ---------- Helpers ----------

    private static async Task AssertSameEndpointAsync(string left, string right)
    {
        static (string Host, int Port) Split(string address)
        {
            var separator = address.LastIndexOf(':');
            return (address[..separator], int.Parse(address[(separator + 1)..]));
        }

        var (leftHost, leftPort) = Split(left);
        var (rightHost, rightPort) = Split(right);
        leftPort.ShouldBe(rightPort, $"Both clients should use the same port, got {left} and {right}");

        var leftIps = await Dns.GetHostAddressesAsync(leftHost);
        var rightIps = await Dns.GetHostAddressesAsync(rightHost);
        leftIps.Intersect(rightIps).ShouldNotBeEmpty(
            $"Both clients should be connected to the same server, got {left} and {right}");
    }

    private string ResolveAddressForRole(string role) => role.ToLowerInvariant() switch
    {
        "leader" => _context.LeaderTcpUrl,
        "follower" => _context.FollowerTcpUrl,
        "single" => _context.TcpUrl,
        _ => throw new ArgumentOutOfRangeException(nameof(role), role, "Unknown role"),
    };

    private string ResolveAddressForPort(int port) => port switch
    {
        8090 => _context.TcpUrl,
        8091 => _context.LeaderTcpUrl,
        8092 => _context.FollowerTcpUrl,
        _ => throw new ArgumentOutOfRangeException(nameof(port), port, "Unknown port"),
    };

    private async Task CreateAndConnectClient(string name, string address)
    {
        var client = IggyClientFactory.CreateClient(new IggyClientConfigurator
        {
            BaseAddress = address,
            Protocol = Protocol.Tcp,
        });
        await client.ConnectAsync();
        _context.Clients[name] = client;
        if (name == "main")
        {
            _context.IggyClient = client;
        }
    }

    private IIggyClient GetClient(string name)
    {
        return _context.Clients.TryGetValue(name, out var c)
            ? c
            : throw new InvalidOperationException($"Client {name} should exist");
    }

    private async Task VerifyClientPort(string clientName, int expectedPort, bool markRedirection)
    {
        var client = GetClient(clientName);
        await VerifyClientConnection(client, expectedPort);

        var leader = await GetLeaderFromMetadata(client);
        if (leader is not null && markRedirection && leader.Endpoints.Tcp == expectedPort)
        {
            _context.RedirectionOccurred = true;
        }
    }

    private static async Task VerifyClientConnection(IIggyClient client, int expectedPort)
    {
        var addr = client.GetCurrentAddress();
        if (!addr.EndsWith($":{expectedPort}"))
        {
            throw new ShouldAssertException(
                $"Expected connection to port {expectedPort}, but connected to: {addr}");
        }
        await client.PingAsync();
    }

    private static async Task<ClusterNode?> GetLeaderFromMetadata(IIggyClient client)
    {
        try
        {
            var metadata = await client.GetClusterMetadataAsync();
            return metadata?.Nodes.FirstOrDefault(n =>
                n.Role == ClusterNodeRole.Leader && n.Status == ClusterNodeStatus.Healthy);
        }
        catch (IggyInvalidStatusCodeException e) when (e.StatusCode == 5)
        {
            return null;
        }
    }
}
